cleanlab/internal/util.py

```diff
@@ -438,14 +401,112 @@ def print_joint_matrix(joint_matrix, round_places=2):
 
 def compress_int_array(int_array, num_possible_values):
     """Compresses dtype of np.array<int> if num_possible_values is small enough."""
-    compressed_type = None
-    if num_possible_values < np.iinfo(np.dtype("int16")).max:
-        compressed_type = "int16"
-    elif num_possible_values < np.iinfo(np.dtype("int32")).max:  # pragma: no cover
-        compressed_type = "int32"  # pragma: no cover
-    if compressed_type is not None:
-        int_array = int_array.astype(compressed_type)
-    return int_array
+    try:
+        compressed_type = None
+        if num_possible_values < np.iinfo(np.dtype("int16")).max:
+            compressed_type = "int16"
+        elif num_possible_values < np.iinfo(np.dtype("int32")).max:  # pragma: no cover
+            compressed_type = "int32"  # pragma: no cover
+        if compressed_type is not None:
+            int_array = int_array.astype(compressed_type)
+        return int_array
+    except:  # int_array may not even be numpy array, keep as is then
+        return int_array
+
+
+def subset_X_y(X, labels, mask):
+    """Extracts subset of features/labels where mask is True"""
+    labels = subset_labels(labels, mask)
+    try:
+        X = X[mask]
+    except:
+        raise TypeError("Data features X must be subsettable with boolean mask: X[mask]")
+    return X, labels
+
+
+def subset_labels(labels, mask):
+    """Extracts subset of labels where mask is True"""
+    try:  # filtering labels as if it is array or DataFrame
+        return labels[mask]
+    except:
+        try:  # filtering labels as if it is list
+            return [l for idx, l in enumerate(labels) if mask[idx]]
+        except:
+            raise TypeError("labels must be 1D np.array, list, or pd.Series.")
+
+
+"""
+try:  # filtering labels as if it were array
+            labels_cleaned = labels[x_mask]
+            labels_subsetted = True
+        except:
+            labels_subsetted = False
+        if not labels_subsetted:
+            try:  # filtering labels as if it were list
+                labels_cleaned = [l for idx, l in enumerate(labels) if x_mask[idx]]
+            except:
+                raise TypeError("labels must be 1D np.array, list, or pd.Series.")
+"""
+
+
+def csr_vstack(a, b):
```

Comment:
if this guarantees sparsity is maintained for sparse matrices, should mention that.

Reply:
added to docstring